### PR TITLE
ddl: Support adding a middle state in altering placement policies

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/21119

How it Works:

ref https://github.com/pingcap/parser/pull/1104

Support middle state during altering placement policies and update `checkPlacementPolicyBeforeCommit` to abort txn in some particular cases.

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: session/session.go [hunk 1]
```diff
@@ -2474,6 +2474,18 @@ func (s *session) checkPlacementPolicyBeforeCommit() error {
 					fmt.Sprintf("table or partition %v's leader location %v is out of txn_scope %v", physicalTableID, dcLocation, txnScope))
 				break
 			}
+			tbl, ok := is.TableByID(physicalTableID)
+			if !ok {
+				err = ddl.ErrInvalidPlacementPolicyCheck.GenWithStackByArgs(
+					fmt.Sprintf("table or partition %v is not found", physicalTableID))
+				break
+			}
+			if tbl.Meta().State == model.StateGlobalTxnWriteOnly {
```
Comment (djshow832) [inline session/session.go:2483]: Is the state set in `TableInfo`? I only see the state is set in the `Job`.

File: ddl/partition.go [hunk 2]
```diff
@@ -1730,20 +1730,30 @@ func onAlterTablePartition(t *meta.Meta, job *model.Job) (int64, error) {
...
+			return ver, errors.Trace(err)
+		}
+		job.SchemaState = model.StateGlobalTxnWriteOnly
+	case model.StateGlobalTxnWriteOnly:
+		// 	global txn write only -> public
+		ver, err = updateSchemaVersion(t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		job.FinishTableJob(model.JobStateDone, model.StatePublic, ver, tblInfo)
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply ddl/partition.go]: addressed, now the tblnfo.State is updated.

**Comment by Yisaer** [comment]
move to https://github.com/pingcap/tidb/pull/21356

**Pull request closed without merging**